app/src/main/java/org/mozilla/tv/firefox/pocket/PocketRepo.kt

```diff
@@ -0,0 +1,106 @@
+    object FetchFailed : PocketRepoState()
+}
+
+/**
+ * Manages backing state for Pocket data, as well as any logic related to
+ * retrieving or storing that data.
+ */
+open class PocketRepo(private val pocketEndpoint: PocketEndpoint) {
+
+    private val mutableState = MutableLiveData<PocketRepoState>()
+    open val state: LiveData<PocketRepoState> = mutableState
+
+    @Suppress("ANNOTATION_TARGETS_NON_EXISTENT_ACCESSOR") // Private properties generate fields so method annotations can't apply.
+    @get:UiThread
+    @set:UiThread
+    private var backgroundUpdates: Job? = null
+    @Volatile private var lastUpdateMillis = -1L
+    private val nextUpdateMillis get() = lastUpdateMillis + CACHE_UPDATE_FREQUENCY_MILLIS
+
+    init {
+        // mutableState.value should always be initialized at the top of init, because we treat
+        // it as non-null
+        @Suppress("SENSELESS_COMPARISON")
+        // Values of BuildConfig can change but the compiler doesn't know that
+        mutableState.value = when {
+            BuildConfig.POCKET_KEY == null -> PocketRepoState.NoKey
+            else -> PocketRepoState.Loading
+        }
+
+        update()
+    }
+
+    fun update() {
+        launch {
+            val fetchedState = requestVideos()
+            updateState(fetchedState)
+        }
+    }
+
+    @UiThread // update backgroundUpdates.
+    fun startBackgroundUpdates() {
+        backgroundUpdates?.cancel(CancellationException("Cancelling unexpectedly active job to ensure only one is running"))
+        backgroundUpdates = startBackgroundUpdatesInner()
+    }
+
+    @UiThread // update backgroundUpdates.
+    fun stopBackgroundUpdates() {
+        backgroundUpdates?.cancel(CancellationException("Stopping background updates"))
+        backgroundUpdates = null
+    }
+
+    private fun startBackgroundUpdatesInner() = launch {
+        while (true) {
+            val delayForMillis = nextUpdateMillis - SystemClock.elapsedRealtime()
+            if (delayForMillis > 0) {
+                delay(delayForMillis, TimeUnit.MILLISECONDS)
+            }
+
+            val newState = requestVideos()
+            // Delay next request if the previous was successful, otherwise try again immediately
+            if (newState is PocketRepoState.LoadComplete) lastUpdateMillis = SystemClock.elapsedRealtime()
+            updateState(newState)
+        }
+    }
+
+    private suspend fun requestVideos(): PocketRepoState {
+        val videos = withContext(DefaultDispatcher) { pocketEndpoint.getRecommendedVideos() }
+
+        return videos?.let { PocketRepoState.LoadComplete(it) } ?: PocketRepoState.FetchFailed
+    }
+
+    private fun updateState(newState: PocketRepoState) {
+        val oldState = mutableState.value!! // State is initialized at the top of init
+
+        val computedState = PocketRepoStateMachine(newState, oldState).computedState()
+
+        if (oldState != computedState) {
+            mutableState.postValue(computedState)
```

Comment:
If a user is looking at the Pocket feed and we do a background update, it's going to update the feed right underneath them, right? Do we handle this case? If we don't, it's going to be a bad user experience. Similar for the mega tile: if a user sees a video they want and it changes, they might be annoyed.

iirc, the previous implementation had a cached version of the feed so it'd never update when the user is looking at it (but, for better or worse, could update if they backed out).

This comment should probably be in the view layer, but curiously, it's only necessary because of an implementation detail (background updates) of the repo. 🤷‍♀️ Then again, this is an important concern for most UIs that used cached data.

Reply:
A few possible solutions.

VM solution
- Update VM to stop listening to updates after a successful one
- Modify mega tile and Pocket screen to use the same VM instance, and make sure it falls out of scope when neither is active.  This would mean retrieving the VM with a `LifecycleOwner` that they share, but that the browser doesn't.  I'm not sure what that would be.

Repo solution
- Pause background updates whenever the menu or the Pocket screen are displayed

There has to be a better solution, both of these are terrible.  I'm thinking about it.

Reply:
Intermediate cache solution
- Write a thin cache that lives between the repo and the VM.  Once a LoadComplete has passed through, stop updating the cache.  Have the ScreenController clear the cache whenever the active screen is not overlay or Pocket screen.

Reply:
Implemented the intermediate cache.  Please take a look and let me know what you think.